### ISSUE TITLE
How to improve performance in yolov8obb training ?

### ISSUE THREAD

**Issue by Abhijeet241093** [ultralytics/ultralytics#21157]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 26/26 [00:02<00:00, 11.87it/s]
                   all        403       2854      0.563      0.603      0.617      0.461
                    LB        126        126      0.833       0.91      0.893      0.553
                    LT        132        132      0.902      0.909      0.927      0.621
                    RB        143        143      0.866      0.916      0.897      0.509
                    RT        147        147      0.915      0.959       0.96      0.817
                blotch          4          4       0.27       0.75      0.526      0.474
                broken         21         82      0.831      0.719      0.742      0.449
                 crack          1          1          0          0          0          0
                  date         87         87      0.907      0.966      0.974      0.938
     date_out_of_label          1          1          0          0      0.142      0.142
          label_broken         38         39      0.651      0.795      0.804      0.569
           label_crack         19         19      0.633      0.842      0.767      0.507
         minor scratch          5         10      0.248        0.3      0.183      0.132
         minor wrinkle         52        191      0.556      0.262      0.293      0.118
               no_date         53         53      0.932      0.981      0.994      0.916
               nolabel         22         22      0.905          1      0.995      0.995
         partial_crack          1          1          0          0          0          0
     partial_date_left          7          7          1      0.659      0.995      0.901
    partial_date_right         10         10      0.521        0.6      0.753      0.657
  partial_label_broken          4          4      0.478          1      0.895      0.406
  partial_no_date_left          4          4          1      0.625      0.888       0.75
 partial_no_date_right          1          1       0.22          1      0.995      0.796
partial_rolling_Sticker         31         41          0          0          0          0
   partial_seal_broken          4          4      0.188       0.25       0.25      0.199
               pinhole          7         12      0.505      0.417      0.298      0.191
               process         17         19      0.163      0.211      0.177      0.116
       rolling_sticker        104        358        0.6      0.483      0.499      0.241
               scratch         51        342      0.764      0.729      0.715      0.436
           seal_broken         44         44      0.823      0.886      0.921      0.858
          text_scratch          6          8      0.562        0.5      0.583      0.365
               wrinkle        155        942      0.607      0.411      0.453      0.177
Speed: 0.0ms preprocess, 1.9ms inference, 0.0ms loss, 0.9ms postprocess per image

Issue: Class imbalance

In this case, how can I improve the detection performance for each individual class? Would reducing the number of classes help improve overall accuracy?

model = YOLO("yolov8x-obb.pt")  # Make sure you're using the correct model name and file

model.train(
    data="data.yaml",
    epochs=1200,
    imgsz=640,
    patience=150,
    # batch=-1  # Optional, comment or define properly if needed
)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Abhijeet241093, thank you for reaching out and sharing your results with Ultralytics 🚀! We're glad to see you're experimenting with YOLO OBB models and seeking ways to improve per-class performance.

Since your question relates to class imbalance and training optimization, we recommend checking out our [Docs](https://docs.ultralytics.com/), specifically the [model training tips](https://docs.ultralytics.com/guides/model-training-tips/), which offer strategies for handling class imbalance and improving underperforming classes.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

For training-related ❓ questions like this, please include as much detail as possible, such as:
- Data distribution and sample images for problematic classes
- Training logs and configuration details
- Any augmentation or sampling strategies already in use

You can also connect with the Ultralytics community for more ideas and real-time support:  
- [Discord](https://discord.com/invite/ultralytics) 🎧  
- [Discourse](https://community.ultralytics.com/)  
- [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Ensure you're running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help guide you efficiently. An Ultralytics engineer will review your issue and assist you further soon! 💡

**Comment by Y-T-G** [comment]
You can try weighted dataloader 

https://y-t-g.github.io/tutorials/yolo-class-balancing/

**Comment by Abhijeet241093** [comment]
Thank you very much. Apart form that can you suggest, other techniques or parameters, I should consider...

**Comment by glenn-jocher** [comment]
For class imbalance issues like yours, consider these additional techniques beyond weighted dataloader:

**Data Augmentation**: Use aggressive [augmentation strategies](https://docs.ultralytics.com/guides/yolo-data-augmentation/) particularly for minority classes like 'crack' and 'partial_crack' - try `mosaic=1.0`, `mixup=0.15`, and custom augmentation pipelines.

**Hyperparameter Tuning**: Experiment with `cls=0.5` (increase classification loss weight), `box=7.5` (for OBB), and lower confidence thresholds during validation. The [hyperparameter tuning guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/) provides systematic approaches.

**Class Consolidation**: Consider merging similar classes (e.g., combine 'crack' and 'partial_crack', or different 'partial_date' variants) to increase sample sizes per class.

**Multi-scale Training**: Use `imgsz=[480, 640, 800]` for varied input sizes, especially helpful for small defects like pinholes.

**Focal Loss**: Add `fl_gamma=2.0` to your training arguments to help with the extreme class imbalance.
